### PR TITLE
[Implement] Purchase in app

### PR DESCRIPTION
**Pull request by ven1984**
- Fetch the items set on iTunes connect
- Implement the purchase
- Implement the restore

### CODE REVIEW

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 1]
```diff
@@ -7,13 +7,106 @@
...
 import Foundation
+import StoreKit
+import RxCocoa
+import RxSwift
 
-final class PurchaseService: PurchaseServicing {
-    func purchase(fail: (PurchaseError) -> Void, complete: () -> Void) {
-        // TODO: Implement
+final class PurchaseService: NSObject, PurchaseServicing, SKProductsRequestDelegate, SKPaymentTransactionObserver {
+    fileprivate var productIDs: Array<String> = []
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: we don't need fileprivate here, use private here is better

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 2]
```diff
@@ -7,13 +7,106 @@
...
 
 import Foundation
+import StoreKit
+import RxCocoa
+import RxSwift
 
-final class PurchaseService: PurchaseServicing {
-    func purchase(fail: (PurchaseError) -> Void, complete: () -> Void) {
-        // TODO: Implement
+final class PurchaseService: NSObject, PurchaseServicing, SKProductsRequestDelegate, SKPaymentTransactionObserver {
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: optional: move delegate implement to extension

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift (hunk 1 above)
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: optional: [String] (To follow the coding guide)

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 3]
```diff
@@ -7,13 +7,106 @@
...
+        }
+    }
+    
+    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
+        if response.products.count != 0 {
+            var products:[SKProduct] = []
+            for product in response.products {
+                products.append(product)
+            }
+            print(products)
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: Why we need print here?

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 4]
```diff
@@ -7,13 +7,106 @@
...
+            productsRelay.accept(products)
+        } else {
+            productsRelay.accept([])
+            print("No products available now")
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: Is this a kind of error?

File: PurchaseApp/Views/Home/HomeViewController.swift [hunk 5]
```diff
@@ -52,11 +55,14 @@ final class HomeViewController: UIViewController {
         super.viewDidLoad()
         setupUI()
         bindViewModel()
+        viewModel.fetchPurchaseItems()
```
Comment (CalvinChangCC) [inline PurchaseApp/Views/Home/HomeViewController.swift]: Load internet event at view will appear or did appear?

File: PurchaseApp/Views/Home/HomeViewController.swift [hunk 6]
```diff
@@ -68,24 +74,49 @@ final class HomeViewController: UIViewController {
...
+                    break
+                case .none:
+                    self.statusImageView.image = UIImage(named: "premium")
+                    self.descriptionLabel.text = self.textProvider.normal
+                    self.advancedButton.isHidden = false
+                    self.intermediateButton.isHidden = false
+                    self.restoreButton.isHidden = false
+                    break
+                case .error(_):
+                    // TODO: After purchase state finish, handle the error
```
Comment (CalvinChangCC) [inline PurchaseApp/Views/Home/HomeViewController.swift]: remove TODO if finished?

File: PurchaseApp/ViewModels/Home/HomeViewModel.swift [hunk 7]
```diff
@@ -18,20 +18,61 @@ final class HomeViewModel: HomeViewModeling {
...
+        context.purchaseService.productsRelay
+            .subscribe(onNext: { [weak self] (products) in
+                guard let self = self else { return }
+                if (products.count > 0) {
+                    self.purchaseStateRelay.accept(.none)
+                } else {
+                    self.purchaseStateRelay.accept(.notSupport)
+                }
+            })
+            .disposed(by: disposeBag)
```
Comment (CalvinChangCC) [inline PurchaseApp/ViewModels/Home/HomeViewModel.swift]: I will use map and bindTo, for example

context.purchaseService.productsRelay
.map { return $0.count > 0 ? PurchaseState.none : PurchaseState.notSupport }
.bind(to: purchaseStateRelay)
.disposed(by: disposeBag)

File: PurchaseApp/ViewModels/Home/HomeViewModel.swift [hunk 8]
```diff
@@ -18,20 +18,61 @@ final class HomeViewModel: HomeViewModeling {
...
+                        break
+                    case .advanced:
+                        self.purchaseStateRelay.accept(.advanced)
+                        break
+                    case .intermediate:
+                        self.purchaseStateRelay.accept(.intermediate)
+                        break
+                }
+            })
+            .disposed(by: disposeBag)
```
Comment (CalvinChangCC) [inline PurchaseApp/ViewModels/Home/HomeViewModel.swift]: The same above

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 9]
```diff
@@ -7,13 +7,106 @@
...
+            case .intermediate:
+                return productsRelay.value[1]
+            default:
+                return nil
+        }
+    }
+    
+    func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
+        for transaction in transactions {
+            if transaction.transactionState == .purchased {
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: Why don't use switch here?

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 10]
```diff
@@ -7,13 +7,106 @@
...
+    func typeForProduct(_ type: PurchaseProductType) -> SKProduct? {
+        guard productsRelay.value.count > 1 else {
+            return nil
+        }
+        
+        switch type {
+            case .advanced:
+                return productsRelay.value[0]
+            case .intermediate:
+                return productsRelay.value[1]
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: We can make sure the order is always correct?

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 11]
```diff
@@ -7,13 +7,106 @@
...
+    
+    func purchase(type: PurchaseProductType) {
+        guard let product = typeForProduct(type) else {
+            return
+        }
+        let payment = SKPayment(product: product)
+        SKPaymentQueue.default().add(payment)
+    }
+    
+    func typeForProduct(_ type: PurchaseProductType) -> SKProduct? {
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: private?

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 12]
```diff
@@ -7,13 +7,106 @@
...
+    func requestProductInfos() {
+        if SKPaymentQueue.canMakePayments() {
+            let productIdentifiers = NSSet(array: productIDs)
+            let productRequest = SKProductsRequest(productIdentifiers: productIdentifiers as! Set<String>)
+    
+            productRequest.delegate = self
+            productRequest.start()
+        }
+        else {
+            print("RequestProductInfos failed")
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: We don't handle this error?

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 13]
```diff
@@ -7,13 +7,106 @@
...
+        }
+    }
+    
+    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
+        if response.products.count != 0 {
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: isEmpty

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 14]
```diff
@@ -7,13 +7,106 @@
...
+                return productsRelay.value[0]
+            case .intermediate:
+                return productsRelay.value[1]
+            default:
+                return nil
+        }
+    }
+    
+    func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
+        for transaction in transactions {
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: Optional: foreach

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 15]
```diff
@@ -7,13 +7,106 @@
...
-        // TODO: Implement
+final class PurchaseService: NSObject, PurchaseServicing, SKProductsRequestDelegate, SKPaymentTransactionObserver {
+    fileprivate var productIDs: Array<String> = []
+    let productsRelay = BehaviorRelay<[SKProduct]>(value: [])
+    var purchaseRelay = BehaviorRelay<PurchaseProductType>(value: .none)
+    var errorSubject = PublishSubject<PurchaseError>()
+    
+    override init() {
+        super.init()
+        productIDs.append("item_id_from_apple_01")
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: Optional: Define product ID in enum will much clear

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift [hunk 16]
```diff
@@ -7,13 +7,106 @@
...
+                errorSubject.onNext(.fail)
+            } else if transaction.transactionState == .restored {
+                SKPaymentQueue.default().finishTransaction(transaction)
+                completeTransaction(transaction)
+            }
+        }
+    }
+    
+    func completeTransaction(_ transaction: SKPaymentTransaction) {
+        if transaction.payment == productsRelay.value[0] {
```
Comment (CalvinChangCC) [inline PurchaseApp/Models/PurchaseService/PurchaseService.swift]: compare with product identifier which is defined in enum will more readable

File: PurchaseApp/Models/PurchaseService/PurchaseService.swift (hunk 4 above)
Comment (ven1984) [reply PurchaseApp/Models/PurchaseService/PurchaseService.swift]: Not sure

**Pull request merged**